src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,190 @@
+             call getLinearizedData() if you need them oredered from 
+             oldest->newest
+        * This direct access method is fast.
+        */
+        const std::vector<T>& getData() const;
+
+        /** linearize method for the internal buffer; this is slower than 
+            the pure getData() but ensures that the data are ordered (oldest at
+             the beginning, newest at the end of the vector
+         * This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+         * :return new linearized vector
+        */
+        std::vector<T> getLinearizedData() const;
+
+        bool operator==(const SlidingWindow& r2) const;
+        bool operator!=(const SlidingWindow& r2) const;
+        T& operator[](UInt index);
+        const T& operator[](UInt index) const;
+
+
+      private:
+        std::vector<T> buffer_;
+        UInt idxNext_;
+}; 
+}} //end ns
+
+/// IMPLEMENTATION
+#include <cmath>
+
+using nupic::util::SlidingWindow;
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(nupic::UInt maxCapacity) :
+  maxCapacity(maxCapacity) {
+  NTA_CHECK(maxCapacity > 0);
+  buffer_.reserve(maxCapacity);
+  idxNext_ = 0;
+}
+
+
+template<class T>
+size_t SlidingWindow<T>::size() const {
+  NTA_ASSERT(buffer_.size() <= maxCapacity);
+  return buffer_.size();
+}
+
+
+template<class T>
+void SlidingWindow<T>::append(T newValue) {
+  if(size() < maxCapacity) {
+    buffer_.emplace_back(newValue);
+  } else {
+  buffer_[idxNext_] = newValue;
+  }
+  //the assignment must be out of the [] above, so not [idxNext_++%maxCap],
+  // because we want to store the value %maxCap, not only ++
+  idxNext_ = ++idxNext_ %maxCapacity;
+}
+
+
+template<class T>
+T SlidingWindow<T>::append(T newValue, bool& isValid) {
+  //handle case of empty buffer (access buff[0]), otherwise return oldest elem
+  T old = buffer_.empty() ? newValue : buffer_[idxNext_]; 
+  //only in this case we drop oldest; this happens always after 
+  //first maxCap steps ; must be checked before append()
+  isValid = (buffer_.size()==maxCapacity); 
+  append(newValue);
+  return old; 
+}
+
+
+template<class T>
+const std::vector<T>& SlidingWindow<T>::getData() const {
+  return buffer_; //may contain trailing "zeros"
+}
+
+
+template<class T>
+std::vector<T> SlidingWindow<T>::getLinearizedData() const {
+  std::vector<T> lin;
+  lin.reserve(buffer_.size());
+
+  //insert the "older" part at the beginning
+  lin.insert(std::begin(lin), std::begin(buffer_) + idxNext_, std::end(buffer_));
+  //append the "newer" part to the end of the constructed vect
+  lin.insert(std::end(lin), std::begin(buffer_), std::begin(buffer_) + idxNext_);
+ return lin;
+}
+
+
+template<class T>
+bool SlidingWindow<T>::operator==(const SlidingWindow& r2) const {
+  return ((this->size() == r2.size()) && (this->maxCapacity == r2.maxCapacity) && 
+    (this->getData()== r2.getData()) ); 
+  //FIXME review the ==, on my machine it randomly passes/fails the test!
```

Comment:
> A better == operator would iterate through both SlidingWindows via the [] 

Doesn't the equals for vectors (internal type of buffer_) already check all the elements are the same? I can do the explicit check of all elements if that does not show a performance slowdown(?)